### PR TITLE
http2: simplifies CodecImpl by implementing http2::adapter::DataFrameSource directly

### PR DESCRIPTION
**Pull request by birenroy**
Signed-off-by: Biren Roy <birenroy@google.com>

Commit Message: Removes nghttp2_data_provider from CodecImpl. Cleanup related to https://github.com/envoyproxy/envoy/issues/23596.
Additional Description:
Risk Level: low
Testing: ran unit test and integration tests locally
Docs Changes:
Release Notes:
Platform Specific Features:

### CODE REVIEW

**Review by bencebeky — DISMISSED** [review]

**Review by bencebeky — APPROVED** [review]
Still LGTM.

File: source/common/http/http2/codec_impl.h [hunk 1]
```diff
@@ -400,6 +398,20 @@ class ConnectionImpl : public virtual Connection,
...
   };
 
+  // Encapsulates the logic for sending DATA frames on a given stream.
+  class StreamDataFrameSource : public http2::adapter::DataFrameSource {
+  public:
+    explicit StreamDataFrameSource(StreamImpl& stream) : stream_(stream) {}
+
+    std::pair<int64_t, bool> SelectPayloadLength(size_t max_length) override;
+    bool Send(absl::string_view frame_header, size_t payload_length) override;
+    bool send_fin() const override { return send_fin_; }
```
Comment (KBaichoo) [inline source/common/http/http2/codec_impl.h:408]: method name not consistent with existing style

File: source/common/http/http2/codec_impl.h [hunk 2]
```diff
@@ -400,6 +398,20 @@ class ConnectionImpl : public virtual Connection,
     void grantPeerAdditionalStreamWindow();
   };
 
+  // Encapsulates the logic for sending DATA frames on a given stream.
+  class StreamDataFrameSource : public http2::adapter::DataFrameSource {
+  public:
+    explicit StreamDataFrameSource(StreamImpl& stream) : stream_(stream) {}
+
+    std::pair<int64_t, bool> SelectPayloadLength(size_t max_length) override;
```
Comment (KBaichoo) [inline source/common/http/http2/codec_impl.h:406]: can these be documented?

File: source/common/http/http2/codec_impl.cc [hunk 3]
```diff
@@ -1644,15 +1610,6 @@ ConnectionImpl::Http2Callbacks::Http2Callbacks() {
         return static_cast<ConnectionImpl*>(user_data)->onSend(data, length);
       });
 
-  nghttp2_session_callbacks_set_send_data_callback(
```
Comment (KBaichoo) [inline source/common/http/http2/codec_impl.cc:1647]: can you remind me why we no longer need this callback for nghttp2?  And the implications of removing DATA_FLAG_NO_COPY ?

**Review by KBaichoo — COMMENTED** [review]
Thank you for your contribution

File: source/common/http/http2/codec_impl.h (hunk 1 above)
Comment (birenroy) [reply source/common/http/http2/codec_impl.h:412]: It's an implementation of an interface from QUICHE code, so changing the method name is not possible in this PR.

File: source/common/http/http2/codec_impl.cc (hunk 3 above)
Comment (birenroy) [reply source/common/http/http2/codec_impl.cc:1647]: This callback became unused when we default-enabled the new QUICHE HTTP/2 codec wrapper API.

The wrapper uses a http2::adapter::DataFrameSource implementation when sending HTTP/2 DATA frames. This PR is removing the extra layers of translation between the nghttp2_data_provider and DataFrameSource data types. The actual DATA frame payload bytes still do not pass through the codec library, so no extra copies are introduced by dropping DATA_FLAG_NO_COPY.

**Review by diannahu — APPROVED** [review]

**Review by KBaichoo — APPROVED** [review]
Thank you for the explanation looks good!

**Pull request merged**